### ISSUE TITLE
Send ArgoCD application errors to Sentry SDN

### ISSUE THREAD

**Issue by jqhuangatflex** [argoproj/argo-cd#13982]
# Summary

The concrete implementation is debatable. My proposed solution is to introduce an additional service type to support scraping ArgoCD internal server errors and exceptions, and make it available to a given Sentry SDN (Source Data Name).

# Motivation
Sentry is one of the most popular application monitoring tool used in the CNCF world. Offering an integration entry point would be a win for both ArgoCD and organizations weighting on Sentry for deployment performance and monitoring.
Here's my use case, I have a Sentry SDN assigned to me, and I want to have all application errors, performance indicators from within ArgoCD architecture for observability purpose.

# Proposal

1. A new service type, with annotation open for configuration to plug in custom SDN;
2. In API server, repository server and application server, flush application errors to the given Sentry SDN, ref: [Sentry-GO-SDK](https://docs.sentry.io/platforms/go/?original_referrer=https%3A%2F%2Fwww.google.com%2F);
3. From user's perspective, we install the new service type alongside with ArgoCD architecture;

**Comment by jqhuangatflex** [comment]
Fix typo: SDN --> DSN (Data Source Name)

**Comment by tonyo** [comment]
(disclaimer: I'm co-maintaining [Sentry Go SDK](https://github.com/getsentry/sentry-go/))

@jqhuangatflex to clarify, by application errors you mean actual Go errors/panics that happen inside ArgoCD components, correct? Sentry SDK integrated in ArgoCD would help here indeed, but I'm also curious: are you hitting too many of those errors?

Regarding performance indicators -- what exactly would you like to track?
It looks like ArgoCD already emits some [metrics](https://argo-cd.readthedocs.io/en/stable/operator-manual/metrics/), and also has support for [OpenTelemetry](https://github.com/argoproj/argo-cd/issues/4972). 

Also, for additional context, if you want to track ArgoCD non-internals event in your jobs (e.g. a `SyncFail`): it looks like one could use [resource hooks](https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/) and, for example, [sentry-cli](https://docs.sentry.io/product/cli/) inside the hook to send a Sentry event.

**Comment by jqhuangatflex** [comment]
> (disclaimer: I'm co-maintaining [Sentry Go SDK](https://github.com/getsentry/sentry-go/))
> 
> @jqhuangatflex to clarify, by application errors you mean actual Go errors/panics that happen inside ArgoCD components, correct? Sentry SDK integrated in ArgoCD would help here indeed, but I'm also curious: are you hitting a lot of those errors?
> 
@tonyo : yes, errors inside ArgoCD components; Not ArgoCD applications errors. We don't know how often do we hit it but I guess not often; This is a new requirement so we don't know clearly what it will bring us.

> Regarding performance indicators -- what exactly would you like to track? It looks like ArgoCD already emits some [metrics](https://argo-cd.readthedocs.io/en/stable/operator-manual/metrics/), and also has support for [OpenTelemetry](https://github.com/argoproj/argo-cd/issues/4972).
We have integrated with ArgoCD metrics for monitoring purpose.

> Also, for additional context, if you want to track ArgoCD non-internal events in your jobs (e.g. a `SyncFail`): it looks like one could use [resource hooks](https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/) and, for example, [sentry-cli](https://docs.sentry.io/product/cli/) inside the hook to send a Sentry event.
This looks promising in sending application sync errors to Sentry. We would like to give it a try. Maybe out of context of this particular request, but @tonyo do you know how to scrape deployment events from the hook? I am trying to use syntax like `sentry-cli send-event -m "Demo Event" --logfile output.log` to send these error logs.

**Comment by tonyo** [comment]
> do you know how to scrape deployment events from the hook?

Haven't worked with ArgoCD much, but since you can run arbitrary containers/code in your hook, you could probably use [`argocd` CLI ](https://argo-cd.readthedocs.io/en/stable/cli_installation/)or `kubectl` to fetch any data you need (such as events).

**Comment by eugenberend** [comment]
I would like to see ArgoCD errors in Sentry. 
Use case: we have suffered from tons of errors during argocd app syncing. We've impemented a lot of fixes in ArgoCD components config and, in addition, upgraded ArgoCD to the latest version. To check if errors were gone, we should collect logs from argocd pods and review it manually using ELK or Loki. We should also remember that new ArgoCD version can also be bugged, just like every other product or tool. Collecting errors or stack traces could be useful and also could help to improve stability and quality of our internal ArgoCD-as-a-service at all.

**Comment by ebuildy** [comment]
This is a good idea! 

We managed to send some events to elasticsearch, by using notifications HTTP webhook https://argocd-notifications.readthedocs.io/en/stable/services/webhook/ but this is clearly not solid.

Will be happy to plug argocd to APM too (sentry or elastic)
